### DISCUSSION TITLE
Slim style FDD needed

### DISCUSSION THREAD

**Discussion by bstajcar** [acemielektron/fddEMU#26] [General]
Just a question: I would like to build a low profile (13-15mm) FFD for mu Toshiba T1850 (and for Agilent scope as well) because they are both faulty (mechanics.. transmission rubber belts are de-composed) so obviously the display must also be low profile to fit into the slot. 
So, how to change display? (I am not very good at coding, sorry.. )
And, is there any hope the source can be compiled in the usual Arduino-way?
Thank you in advance!
Best regards, 
Bojan

**Comment by acemielektron** [comment]
Thank you for your message bstajcar,

fddEMU might work with the slimmer 128x32 i2c oled screen after some
adjustments in code.
If possible first check fddEMU works with your laptop and scope, if it
works please let me know, I will try to code for smaller screen.
Arduino IDE threw errors compiling fddEMU last time I tried, which was over
a year ago. I don't know if things changed but I wouldn't count on it.

acemielektron

**Reply by bstajcar** [reply]
OK, thank you for your quick reply :-)
I ordered connector adapters and extenders  (original FDD has 26pin flex cable) so my test will take some time, but as soon as I have it working on protoboard, I will let you know. 
Regards, 
Bojan

**Comment by bstajcar** [comment]
Hi acemielektron, 
Finally I found the time to continue with this project..
Upload went fine with Xloader (debug version) and I was able to see on terminal screen the f/w works OK.
H/w will look like below (not wired yet, I want to finish this and test on comp in couple of days), it will be slim profile and hopefully will fit into Agilent CRO  tray.

Now I have a question: Is there a quick/easy way to format SD card (I have 2GB.. that means there will be ~2k of files on it) ? 

Thank you in advance and best regards, 
Bojan 

![c_76](https://github.com/acemielektron/fddEMU/assets/17623823/4e306658-8b7a-4b3f-9ebf-fa849d815d96)
![c_75](https://github.com/acemielektron/fddEMU/assets/17623823/a0e8a525-fb63-479e-8063-3457f5f675c2)

**Reply by bstajcar** [reply]
OK, I found it! 
USB_Floppy_ManagerII_V1.31T

**Comment by bstajcar** [comment]
Hi acemielektron,
It seems it works OK.. 
I decided to  use [5-way switch](https://www.ebay.com.au/itm/382851250934?hash=item5923b67af6:g:gIoAAOSw8kxilcbg&amdata=enc%3AAQAIAAAA4G3NT5FKSMgJeofwk08pYiJUVaLhDhXd9RyXl1qYJy7iNgsIpG6z6MRXeNGX3OeBLuZEq0cNFxaINKnw3P3oKoSp5Sr7%2Bn4Ch1ukOvxVl6AQFe0SLMEeqaZsPvMF1MdYjc9PqpbAzsrMx1xo3AMOJhHfm4ozS5%2BaYTm%2FgV4N2U0D%2FvRUaCjs%2BysAXf3zLiEzBtMQVYvw0AabD98SdYNBTfSkaiqhrp9pi4M1J0xjp2m0MHVThum3idWyfoPQBol6iVGQTstyWdHj6o4WflboMPU1bb0V1ilZ1%2BiYkkOky0xY%7Ctkp%3ABFBM1qPD1KJi) instead of 5 push-buttons in my design, to make room for LCD at the front. 
The design the PCB will be similar to HxC, with 3 different interface connectors (flat ribbon and IDC connectors).
So perhaps now is the time to modify the code to accommodate low profile LCD. 
Can you help me with this?  I would greatly appreciate if you can do it, since am not very good  at writing code.
Thanks in advance!
Bojan
![c_75](https://github.com/acemielektron/fddEMU/assets/17623823/85f4306d-fa67-4d5a-9aa4-82dccc22cd90)
![c_77](https://github.com/acemielektron/fddEMU/assets/17623823/52e7474d-2f34-4a2e-8ad5-e535236c92ac)

**Comment by acemielektron** [comment]
Thanks for writing.
I have been planning to adapt fddEMU to 128x32 OLED for a while. I just
couldn't find time for it. Now with your request, I hope I can adapt the
code soon.

**Comment by bstajcar** [comment]
Thank you!
And I will certainly share the PCB design with  you when I finish it..
I am thinking about making the board a bit smaller than actual micro-FFD form factor.. so the enclosure (or just front plate) that fits into the device space can be 3D printed.

**Comment by acemielektron** [comment]
Try make HALFLCD=1 FLIP=0 DUAL=0. or try https://github.com/acemielektron/fddEMU/releases/tag/23e73d3
initializes LCD as 128x32, suitable for single drive config. Floppy image loading
works. Splash and Info/Warning screens needs redesign.
Also your 5 way switch idea is very nice. Resulting build would be really compact.

**Comment by bstajcar** [comment]
Excellent! 
I will try asap (today) and report.

**Comment by bstajcar** [comment]
I have a problem now...
I formatted 2GB SD, and placed *.IMG (created with WinImage and renamed from IMA to IMG).
I can see directory on LCD screen
However fddEMU sometimes says "No disk". what would be the cause for this? 
How to make image file contiguous (if that is the cause of this behavior)?

LCD display and terminal screen are not quite consistent and a bit confusing... Which button corresponds to what command from keyboard?
It seems it works OK when I use keyboard commands...
And finally, do I have to have ffdEMU connected to PC via floppy drive cable?


![c_79](https://github.com/acemielektron/fddEMU/assets/17623823/0dc34e28-6008-439d-8001-422f24a0b86c)
![c_87](https://github.com/acemielektron/fddEMU/assets/17623823/64ec9ec2-c98e-4217-a92a-a999bc64985c)

**Comment by acemielektron** [comment]
No disc means, no floppy image loaded. You can either load a floppy image
from the SD card using screen and buttons or using the serial terminal. A
good alternative while testing would be placing a floppy image named
"boot.img" on SD card which fddEMU auto loads at startup.
If (despite placing floppy images on SD) there is no floppy image listed on
either screen menu or serial terminal this means either fddEMU can't read
sd-card due to broken electrical connections or wrong filesystem.
Just make sure you have formatted your SD card as fat16 or fat32 then place
floppy images (as many as you want) on it.

**Comment by bstajcar** [comment]
OK, it seems this is OK now. 
I didn't read the instructions properly (happens to me very often!), but now I think understand the procedure better.
Files are listed on LCD screen, but not on terminal (via USB).
Upon reset, terminal screen looks like below. 
First, there is "ERROR: Not found" then after couple of seconds "Select BOOT1" (the image file of system disk I renamed as "BOOT1.IMG"
Command "L" from keyboard does no effect (? maybe I need to have connection to PC at this stage?)
So the next step is connecting to PC and try to load files via floppy connection as from real floppy.
![c_88](https://github.com/acemielektron/fddEMU/assets/17623823/4f2bfee2-f9aa-4501-aa11-f6fcd57d2199)

**Comment by bstajcar** [comment]
Another hurdle (I hope the last one :-) ) I connected the fddEMU to desktop via USB-FDD adapter.. 
w'7 wants to format the disk... 

![c_91](https://github.com/acemielektron/fddEMU/assets/17623823/ce1f30d6-c03e-4fc5-b0bb-5d4c0ebd7f54)

![c_92](https://github.com/acemielektron/fddEMU/assets/17623823/e5757209-6b6b-42a0-8ac3-9a3de4897ca0)

**Comment by bstajcar** [comment]
Front panel PCB
![c_94](https://github.com/acemielektron/fddEMU/assets/17623823/b9ac4d6f-1a2b-42da-a887-d152d95ff6ae)

**Comment by bstajcar** [comment]
Hi Acemi,
Still no joy... From LCD point of view everything looks OK: disk image
is recognized, on terminal I can even see info on floppy (below, it is
1.44MB DOS6.22 with some apps on it)

********
FS: FAT12
BPS: 512 Heads: 2 SPT: 18 Total Sec: 2880
BUSY
A:R0/0/1
A:R0/0/2
A:R0/0/3
A:R0/0/4
A:R0/0/5
A:R0/0/6
A:R0/0/7**********When tried to access the drive, I can see on LCD
screen the icon (presumably "card active", and terminal shows activity
but, PC can't read the drive and wants to format it (I didn't try this
because I didn't want to introduce more unknowns in my testing.I
triple check the wiring between IDC and NANO, it's correct....
So... what  else can I do? 
The only suspect now is that USB-floppy interface, I will try it with
"real" FDD.. And of course, I may want to use PC with FDD inteface..
Thanks for your help and best regards, 

Bojan

----- Original Message -----

**Comment by acemielektron** [comment]
Hi Bojan,
I am not sure why image is not recognized. I will test fddEMU and let you
know if something got broken.

**Comment by bstajcar** [comment]
Thank you! 
I also tested 2 mechanical FDs.. Interestingly, this USB-FDD adapter (from aliexpress) seems to recognize only 1.44 floppies.. when I insert 720kb, windows behaves in the same manner as with fddEMU... so it could be some in-compatibility there. I have to find old PC with FDcontroller to try again.   
In the meantime, I was working on PCB design... finished, need to send it to PCBFast.


![c_78](https://github.com/acemielektron/fddEMU/assets/17623823/78582ce1-d244-47f1-9bb2-2eaee622a35d)
![c_77](https://github.com/acemielektron/fddEMU/assets/17623823/857a9b7d-0227-4fd3-af44-f5efbe7713c9)

**Comment by acemielektron** [comment]
Sorry for the late reply. I had to rebuild a testing rig (arduinoFDC). Then
repair my fddEMU, without a PCB cabling is tricky. I have tested reading
with both arduinoFDC and a PC. The result: ok, both tests passed. Since I
haven't changed anything in emulation code, this was expected.
So if things doesn't work, please test cabling, and the 2 pull-up
resistors.
Please let me know your results.
Thanks

**Comment by bstajcar** [comment]
Thank you for the reply :-)And thank you for your confirmation effort.
I was expecting all to be OK from your side, but now I feel much more
confident for the good outcome here.
At this moment, I am building arduinoFDC in my spare time, waiting for
crimping tool to make cabling between ArduinoFDC (A-NANO)  and fddEMU
properly  (I couldn't find working PC), so I will certainly let you
know the result when I have that working.This is going to be a really
good project, it has great potential :-) 
I hope my PCB will also turn out good.
  
Best regards from Down under, 
Bojan

----- Original Message -----

**Comment by bstajcar** [comment]
I still have a problem..

On ArduinoFDC terminal plot, 
Top part is fddEMU response after power up, the lower part is  ArduinoFDC (Error #3).

However, when I change ArduinoFDC from A: to B:, it works... 
So I moved jumpers as per second image and now fddEMU works OK with ArduinoFDC.
However,  it still doesn't work from PC (via USB_FDC). 

(I used "atmega328p-single-debug.hex").

![c_81](https://github.com/acemielektron/fddEMU/assets/17623823/9c8dd102-1763-4f53-b502-f783328154a2)


![c_84](https://github.com/acemielektron/fddEMU/assets/17623823/35f1a87c-649e-4312-9430-6ed82043a2dc)

**Comment by bstajcar** [comment]
Hi acemielektron, 
I was wandering..  
fddEMU definetly works OK via ArduinoFDC interface, and most probably  all will be OK on PC (still trying to find one for test).
When testing and investigating problem with Aliexpress [USB 1.44 adapter](https://www.aliexpress.com/item/4000293262310.html?spm=a2g0o.productlist.main.1.5d7238b6WS0rfq&algo_pvid=82574d85-935c-48b6-b25b-550de614ef78&algo_exp_id=82574d85-935c-48b6-b25b-550de614ef78-0&pdp_npi=4%40dis%21AUD%2113.96%2111.45%21%21%218.86%21%21%402101f49716922361061601373e8d9b%2112000026843482389%21sea%21AU%21866261649%21&curPageLogUid=o8Hqy7kQqLct) annd comparing it with ArduinoFDC,  I noticed fddEMU outputs constant 250kHz square wave from D9 once data transmission is done, see osciloscope screen below (lower row. Upper row is the same signal with mechanical FDD)). There is also some jitter visible on D9, not present (or smaller) on "real" (mechanical FDD. 
Could one of those be the reason USB adapter doesn't recognize fddEMU?
![Comparison-FDD vs fddEMU](https://github.com/acemielektron/fddEMU/assets/17623823/b83c26cc-0c71-40b8-be10-a59dafef32c7)

**Comment by acemielektron** [comment]
Hi, Bojan

I wanted to examine fddEMU output on my build before replying but that
required building a somewhat working  ossiloscope which I haven't been able
to, yet. I might have tried those AliExpress USB 1.44 adapters myself but
it's cost (due to tax in my country) is prohibitively expensive for such a
simple device.

From what I understand you have already tested your build with with Arduino
FDC and it worked. When drive select and motor on signals are asserted
fddEMU like a real fdd would output data, this data should be 500khz MFM
for HD (1.44M) disks. And whenever these two signals are deasserted fddEMU
should stop pumping data. Perfect square waves at 250khz means continuous
MFM encoded 0. That is no meaningful data. It shouldn't have worked with
Arduino FDC at all. Even sector data is all zeroes there are sector
headers, gaps, etc. fddEMU shouldn't output continuous zero.
If I could set up an ossiloscope I would look at the traces myself and
compare to yours. In the mean time, looking forward to hearing from you.

acemielektron

**Comment by bstajcar** [comment]
Hi Acemielektron,

Thank you for your quick reply.

Yes, perhaps you could try to borrow oscilloscope from somewhere (old, analogue one is also quite OK for this kind of investigations), because the same waveform is also present when I was using ArduinoFDC, but as you saw on my screen dumps, "DIR" command resulted in OK-looking directory list (despite spurious that continuous 250kHz waveform after first data transfer, which apparently didn't have any ill effect here.. maybe it will be OK in PC as well? I will let you know when I find one for test - they are very rare these days, and I am reluctant to dismantle my Agilent oscilloscope for test at the moment).

Thanks again for your help :-)

Regards, 

Bojan 

BTW, you can probe D9 output without oscilloscope, you can use multimeter (DC voltmeter). 
If the 50% duty cycle waveform is output (as on my setup), your DC voltmeter will show 2.5V  instead of 0V or 5V (LOW or HIGH).

**Comment by acemielektron** [comment]
Ok, I remember now, that continuous zero is by design, I mean I programmed
timer1 to output MFM 0 when idle. I just didn't thought about that before.
You know arduino generates PWM by using timers, atmega328 has two 8 bit
timers (timer0 and timer2) and single 16 bit timer (timer1) which is used
in fddEMU to both generate MFM and capture MFM coming from the controller
(it is atmega328's only timer with capture capability). I thought loading
MFM pattern for 0 is better than loading timer with something random, such
as last loaded value. From floppy drive controller's (FDC) perspective this
is not important, because it is not reading floppy data while idle. If you
have read the readme.md, I am sure you have seen "*fddEMU** is not an cycle
exact floppy emulator , it is more of a hack (it's a bit slower than actual
fdd) but it works because floppy controllers have sensible timeouts and we
take advantage of that*." (first line describing limitations of fddEMU).
Why? Because the 1.44M HD floppy has 18 sectors of 512 bytes on each track,
and  to be faithful to timings we need to cache all of them in RAM. For 18
sectors we need 18x512b = 9Kb of RAM which atmega328 certainly doesn't
have. So we send the first sector, read the second from the sd-card, then
send that. We have no way of automating either reading sector data from sd,
or writing read data to FDC, because atmega328 lacks a DMA controller. On
the other hand FDC doesn't need one sector coming immediately after the
previous one, as long as it reads the requested sector within 5 rotations
of track (200ms each) it would be happy and wont raise an error. So
although the next sector's data comes a little later compared to a real
floppy as long as the data comes within allotted time FDC doesn't care. And
each successful read of a sector resets that timer. FDC also doesn't care
what is coming from the read line unless it is actively reading it, which
it signals by asserting both motor on and drive select signals. So fddEMU
works.
I don't know how they programmed usb floppy adapters sold at aliexpress,
but I doubt it complies with PC FDC behaviour. I guess it is another hack.
fddEMU being a hack itself probably doesn't work with such adapters. I
guess a sensible firmware update on those usb floppy adapters would enable
not only reading fddEMU but also 720K DD disks and other 5.25" formats.
In hindsight Raspberry pi pico (RP2040 has 264K RAM and also has DMA) which
is now even cheaper than arduino nano -which was not available at the time
I have written fddEMU- is a much better choice for implementing a floppy
emulator.

**Comment by bstajcar** [comment]
Yes, I think you are 100% right here, I also think those USB adapters may be problematic (they work in most cases (only with 1.44 floppies), I tried 4 mechanical FDD's and 3 worked OK, one could be faulty...), so I didn't really think that it is actually any indication of fddEMU problem.
No worries, I still intend to send my PCB design to PCBway for prototyping on first occasion.. and in the meantime I am looking for old and suitable PC to confirm all is OK with fddEMU.
Thanks for your help :-)
Regards, 
Bojan

**Comment by acemielektron** [comment]
I think relatively old desktop main boards (10+ years old) might have
floppy drive connector. If you have a desktop computer check the main board
you might be surprised.
